Pull request for adding simple command line based tutorial.

BJackal opened the pull request:
Added two new tutorials for executing Chaste tests with command line arguments through bash scripts. 

The first is a very simple Hello World style example which takes in multiple command line arguments and outputs the sum. For now this has been placed in global as it seemed the most appropriate location.

Another example has been added for running a bash script with command line arguments to execute multiple vertex based simulations with the height and width of tissue based on command line parameters. This is a slightly expanded version of the vertex based simulation tutorial.

BJackal committed:
Simple tutorial to take some command line arguements from a bash script for a test. Add those values then cout the result.
Test needs to be ran from bash script and not directly. Will add some kind of assertion that a command line option has been taken otherwise ERROR out.

BJackal committed:
Cleaned up the tests a little and added some notes to the bash scripts and accompanying test files. Both tests utilise bash scripts and are found one directory above the respective test.

A generic command line arguement example has been added alongside a use case for a vertex mesh.

BJackal committed:
Update TestCommandLineVertexMeshTutorial.hpp

Improving comments.

BJackal committed:
Update runcommandlinemeshtutorial.sh

Adding some extra comments.

BJackal committed:
Update runcommandlinetutorial.sh

Improving commenting.

BJackal committed:
Update TestCommandLineArgumentsTutorial.hpp

Slight editing of comments.

BJackal committed:
Update TestCommandLineVertexMeshTutorial.hpp

Removing null assertion for now as it breaks due to outp not being a pointer. Will be added back in later.

BJackal committed:
Update TestCommandLineVertexMeshTutorial.hpp

Changing test locations with recent changes to Chaste

BJackal committed:
Merge branch 'develop' into 267-command-line-tutorial

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,37 @@
+#ifndef TESTCOMMANDLINEARGUMENTSTUTORIAL_HPP_
```
Needs the copyright notice (this is causing a [test to fail](https://github.com/Chaste/Chaste/actions/runs/9385768547/job/25844727324?pr=282)). Should be a matter of either copying and pasting from another test or running the script mentioned in the failure message.

BJackal replied:
Corrected in https://github.com/Chaste/Chaste/pull/282/commits/3c90aa18a90573a59675f081acadf022734b3a30

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,37 @@
+ * be used to run multiple tests with varying parameter inputs.
+ */
+
+ // NOTE: This test will not work if directlly executed from terminal due to requiring command line arguements.
```
Sounds as if it will make the test pack fail as well then? We can't really be having that, might need to put heads together on how to do a tutorial a different way...

It should be easy to do an "overwriting defaults" using e.g.
```c++
unsigned outp1 = 0;
if CommandLineArguments::Instance()->OptionExists("-opt1")
{
    outp1 = CommandLineArguments::Instance()->GetUnsignedCorrespondingToOption("-opt1");
}
```
and then we can put in comments saying "Please rerun the executable with arguments to see how the output changes, but note you will have to do this outside the `ctest` framework".

BJackal replied:
I can think of one alternative way. The test could have some default conditions that are ran if command line arguments are not detected. Allowing the tutorial to simply run with those default initial conditions.

BJackal replied:
This will similarly be changed for TestCommandLineArgumentsTutorial

BJackal committed:
Adding correct copyright notice.

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,37 @@
+ * be used to run multiple tests with varying parameter inputs.
+ */
+
+ // NOTE: This test will not work if directlly executed from terminal due to requiring command line arguements.
```
```suggestion
 // NOTE: This test will not work if directly executed from terminal due to requiring command line arguments.
```

mirams committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

BJackal committed:
Merge branch '267-command-line-tutorial' of github.com:Chaste/Chaste into 267-command-line-tutorial

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,72 @@
+    void TestCommandLineTutorial()
+    {
+        // Here we utilise CommandLineArguments::Instance()->GetUnsignedCorrespondingToOption() to take in our command line arguements.
+        int outp1 = CommandLineArguments::Instance()->GetUnsignedCorrespondingToOption("-opt1");
```
variable is an `int` but that should return an `unsigned` (int)? Not sure all compilers will like that or whether we should suggest as best practice in a tutorial, I would replace with unsigned variables here probably

BJackal replied:
Edited as suggest

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,72 @@
+class TestCommandLineArgumentsTutorial : public CxxTest::TestSuite
+{
+public:
+    void TestCommandLineTutorial()
```
Probably also want to show off some of the other nice methods in CommandLineArguments. e.g. getting `double`s and also getting vectors is a nice one with e.g.
```
./executable --my-vector-of-arguments 1.1 2.2 3.3 4.4
```
using 
```
std::vector<double> some_doubles = CommandLineArguments::Instance()->GetDoublesCorrespondingToOption("--my-vector-of-arguments");
```

mirams commented on cell_based/test/tutorial/TestCommandLineVertexMeshTutorial.hpp:
```diff
@@ -0,0 +1,211 @@
+/*
```
I'm not really sure we need a whole vertex simulation here, it is bloating things a bit for no real benefit (there's no need for it to be a big complicated example, and nothing cell-based specific about passing in arguments). 

What if we just copy and paste the bash script into the `TestCommandLineArgumentsTutorial.hpp` to be rendered as a block of code on the website for readers to paste into a bash script? To e.g. just run the tutorial with different inputs in a loop?

BJackal replied:
I can see this being very useful. Creating multiple tests within `TestCommandLineArgumentsTutorial.hpp` with a commented bash script for each. Then informing the user to copy and paste each variant of the bash script into their file for each test they wish to try out. Each test could then be showing how to execute with various methods, such as using vectors. As you mentioned this would look much cleaner on the website also.

mirams reviewed (changes requested):
A few suggestions, I think I would simplify quite a bit as there's nothing vertexey or cell based about wanting to pass in arguments, and would complicate the tutorial to show off more of the capabilities of the `CommandLineArguments` class!

BJackal committed:
Removed vertex based command line example.

Modified command line global tutorial to have two tests. First for simple command line arguments and second showing off taking in doubles and utilising a vector
of arguments.

Each test now also checks if command line options have been given and if not will default to a basic set of parameters to ensure the test will run correctly.

Included a paste of the bash script within the test for tutorial purposes.

BJackal committed:
Fixing sum used uninitalised.

BJackal commented:
1. Removed the vertex based simulation and its accompanying bash script.
2. Added a secondary test for the TestCommandLineArgumentsTutorial using a vector of doubles as the input.
3. Added a commented section to each test with the required bash script to run them, by default left the simpler test as the pre-existing bash script.
4. Changed int in the first test to unsigned as suggested.
5. The bash script to take in doubles requires awk (or a similar other tool) to allow for doubles to be input as arguments. 
6. Added a default case for each test if the correct corresponding command line argument is not detected to prevent the test failing.

Possible further additions I can think of now, it may be best to add some simple testing into the default cases or exceptions/assertions to ensure it is running as expected. Additionally , as the CommandLineArguments has the `GetBoolCorrespondingToOption` method should a test just showing this also be added ?

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+    /* The Default test should be ran with the following bash script starting from line 16.
+    runcommandlinetutorial.sh by default has this script already. However, if you have changed
+    the bash script for another text simply copy and paste between START and END back into the file
+    **START**
```
```suggestion
    ```sh
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+          done
+        done
+      done
+    **END**
```
```suggestion
    ```
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+      This could be very useful for several reasons, such as having a large list of parameters one may wish to pass into a test. 
+
+      For this test change the runcommandlinetutorial.sh bash script from line 16 onwards as follow by copy and pasting between START and END.
+      **START**
```
```suggestion
      ```sh
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+        done
+       done
+      done
+      **END**
```
```suggestion
      ```
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+           }
+           // We will finally cout our result for the summed vector components.
+           // Here we have chosen to set our precision to 4 to ensure the correct number of significant figures.
+           std::cout << std::setprecision(4) << "When we add all variables in our vector together we get " << sum  <<"\n"; 
```
```suggestion
           std::cout << std::setprecision(4) << "When we add all variables in our vector together we get " << sum  <<std::endl; 
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+           }
+           // We will finally cout our result for the summed vector components.
+           // Here we have chosen to set our precision to 4 to ensure the correct number of significant figures.
+           std::cout << std::setprecision(5) << "When we add "<< vector_of_doubles[0] << ", " << vector_of_doubles[1] << " and " << vector_of_doubles[2] << " together from our vector we get " << sum  <<"\n"; 
```
```suggestion
           std::cout << std::setprecision(5) << "When we add "<< vector_of_doubles[0] << ", " << vector_of_doubles[1] << " and " << vector_of_doubles[2] << " together from our vector we get " << sum  <<std::endl; 
```

mirams commented on global/runcommandlinetutorial.sh:
can we just remove this one if it is in the generated tutorial?

BJackal replied:
With the code block you suggested instead carrying out this bash script it makes sense to remove it for tidying purposes I think. Its possibly also a good way to set up best practice for such bash scripts if there isn't already one. So there aren't multiple bash scripts floating around.

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+    }
+};
+
+#endif /*TESTHELLO_ECADTURNOVERMODEL_HPP_*/
```
```suggestion
#endif /*TESTCOMMANDLINEARGUMENTSTUTORIAL_HPP_*/
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+           // Here we have chosen to set our precision to 4 to ensure the correct number of significant figures.
+           std::cout << std::setprecision(5) << "When we add "<< vector_of_doubles[0] << ", " << vector_of_doubles[1] << " and " << vector_of_doubles[2] << " together from our vector we get " << sum  <<"\n"; 
+
+        } else {
```
Can we just remove the else block, it is not doing anything here?

BJackal replied:
It could be removed but I think I could also add some basic testing style structure here ? As other tutorials seem to have some small additions for testing purposes. I suppose it could also act as a way to ensure nothing odd happens. I'm also with for removing it though.

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+          for ((k = 2; k <= M; k += 1)); do
+          # Each loop runs an instance of the TestCommandLineArgumentsTutorial with opt1,opt2 and opt3 taking on the 
+          # values of i,j and k resepctivley.      
+          ~/build/global/test/TestCommandLineArgumentsTutorial -opt1 $i -opt2 $j -opt3 $k &
```
Path here is hardcoded, suggest altering instructions to say it should be run "within the build folder". If so will this work?:
```
./global/test/TestCommandLineArgumentsTutorial <args>
```
Test it first(!)

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+
+    /* The Default test should be ran with the following bash script starting from line 16.
+    runcommandlinetutorial.sh by default has this script already. However, if you have changed
+    the bash script for another text simply copy and paste between START and END back into the file
```
```suggestion
    the bash script for another text simply copy and paste the below code into a shell script (`.sh`) file:
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+
+           # Each loop runs an instance of the TestCommandLineArgumentsTutorial with a vector
+           # containing the double corrected version of our varaibles.
+           ~/build/global/test/TestCommandLineArgumentsTutorial --my-vector-of-arguments $idouble $jdouble $kdouble &
```
same as above

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+      In addition to using unsigned integers as command line arguments we can also pass in both doubles and a vector of arguments.
+      This could be very useful for several reasons, such as having a large list of parameters one may wish to pass into a test. 
+
+      For this test change the runcommandlinetutorial.sh bash script from line 16 onwards as follow by copy and pasting between START and END.
```
```suggestion
      To script the running of this test with lots of different arguments, copy and paste the following code into a bash script (`.sh`) file.
```

mirams commented on global/test/TestCommandLineArgumentsTutorial.hpp:
```diff
@@ -0,0 +1,173 @@
+{
+public:
+
+    /* The Default test should be ran with the following bash script starting from line 16.
```
I'd suggest removing bash script and altering this to give tutorial instructions to copy and paste the whole code block below into a new script inside the build folder.

mirams reviewed (changes requested):
A few tweaks @BJackal I think it is getting there, the main one is probably to just get the tutorial to put the bash script in a code block, and give instructions to copy and paste into a bash script in the build folder, then you can remove the script and the need for hardcoded paths I think.

mirams commented:
(you can manually run the generate tutorials and website build to see how the generated webpage looks following these instructions: https://github.com/Chaste/Chaste/issues/193#issuecomment-1908057374)

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested tidying.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Update global/test/TestCommandLineArgumentsTutorial.hpp

Committing suggested edits.

Co-authored-by: Gary Mirams <gary.mirams@gmail.com>

BJackal committed:
Amended pathing to no longer be hard coded and instead intended to be ran from inside the build directory.

Removed bash script within main chaste source code as this should now be created by the user in their build directory.

mirams commented:
Didn't mean to delete branch, we'll do that when ticket is closed!

We'll want to add (manually) an entry to the "Core Functionality" part of the menu page: https://chaste.github.io/docs/user-tutorials/

BJackal commented:
@mirams. Made a branch in Chate.github.io for including the command line arguments tutorial under core functionality and made a pull request.

mirams commented:
OK, closed by https://github.com/Chaste/Chaste.github.io/pull/95
